src/main/java/com/turn/ttorrent/client/Piece.java

```diff
@@ -160,18 +160,27 @@ public synchronized boolean validate() throws IOException {
 		this.valid = false;
 
 		try {
-			// TODO: remove cast to int when large ByteBuffer support is
-			// implemented in Java.
 			ByteBuffer buffer = this._read(0, this.length);
-			byte[] data = new byte[(int)this.length];
-			buffer.get(data);
-			this.valid = Arrays.equals(Torrent.hash(data), this.hash);
+			this.valid = checkHash(buffer);
 		} catch (NoSuchAlgorithmException nsae) {
 			logger.error("{}", nsae);
 		}
 
 		return this.isValid();
 	}
+	
+	protected boolean checkHash(ByteBuffer data) throws NoSuchAlgorithmException {
+		byte[] calculatedHash = Torrent.hash(data);
+		
+		int torrentHashPosition = getIndex() * Torrent.PIECE_HASH_SIZE;
+		for (int i = 0; i < Torrent.PIECE_HASH_SIZE; i++) {
+			byte value = this.torrentHash.get(torrentHashPosition + i);
```

Comment:
Fetching byte per byte on a direct buffer is faster than just fetching the whole thing once ?